spec/javascripts/packages/document-capture/components/document-capture-spec.jsx

```diff
@@ -86,218 +75,6 @@ describe('document-capture/components/document-capture', () => {
     await findByText('doc_auth.errors.camera.blocked_detail');
   });
 
-  it('progresses through steps to completion', async () => {
```

Comment:
I have forgotten why I removed these

Reply:
We shouldn't remove them, at least not the parts which don't affect selfie.

Reply:
```diff
diff --git a/spec/javascripts/packages/document-capture/components/document-capture-spec.jsx b/spec/javascripts/packages/document-capture/components/document-capture-spec.jsx
index eb639a6d5..12001fe54 100644
--- a/spec/javascripts/packages/document-capture/components/document-capture-spec.jsx
+++ b/spec/javascripts/packages/document-capture/components/document-capture-spec.jsx
@@ -1,7 +1,11 @@
 import sinon from 'sinon';
 import userEvent from '@testing-library/user-event';
+import { render as baseRender, fireEvent } from '@testing-library/react';
 import { waitFor } from '@testing-library/dom';
-import httpUpload from '@18f/identity-document-capture/services/upload';
+import httpUpload, {
+  UploadFormEntriesError,
+  toFormEntryError,
+} from '@18f/identity-document-capture/services/upload';
 import {
   ServiceProviderContextProvider,
   UploadContextProvider,
@@ -22,6 +26,10 @@ describe('document-capture/components/document-capture', () => {
   const sandbox = useSandbox();
   const { initialize } = useAcuant();
 
+  function isFormValid(form) {
+    return [...form.querySelectorAll('input')].every((input) => input.checkValidity());
+  }
+
   let originalHash;
   let validUpload;
 
@@ -75,6 +83,182 @@ describe('document-capture/components/document-capture', () => {
     await findByText('doc_auth.errors.camera.blocked_detail');
   });
 
+  it('progresses through steps to completion', async () => {
+    const { getByLabelText, getByText, getAllByText, findAllByText } = render(
+      <DeviceContext.Provider value={{ isMobile: true }}>
+        <AcuantContextProvider sdkSrc="about:blank" cameraSrc="about:blank">
+          <DocumentCapture />
+        </AcuantContextProvider>
+      </DeviceContext.Provider>,
+    );
+
+    initialize();
+    window.AcuantCameraUI.start.callsFake(async (callbacks) => {
+      await Promise.resolve();
+      callbacks.onCaptured();
+      await Promise.resolve();
+      callbacks.onCropped({
+        glare: 70,
+        sharpness: 70,
+        image: {
+          data: validUpload,
+        },
+      });
+    });
+
+    // Attempting to proceed without providing values will trigger error messages.
+    let submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.click(submitButton);
+    const errors = await findAllByText('simple_form.required.text');
+    expect(errors).to.have.lengthOf(2);
+    expect(document.activeElement).to.equal(
+      getByLabelText('doc_auth.headings.document_capture_front'),
+    );
+
+    // Providing values should remove errors progressively.
+    fireEvent.change(getByLabelText('doc_auth.headings.document_capture_front'), {
+      target: {
+        files: [validUpload],
+      },
+    });
+    await waitFor(() => expect(getAllByText('simple_form.required.text')).to.have.lengthOf(1));
+    expect(document.activeElement).to.equal(
+      getByLabelText('doc_auth.headings.document_capture_front'),
+    );
+
+    await userEvent.click(getByLabelText('doc_auth.headings.document_capture_back'));
+
+    // Continue only once all errors have been removed.
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+    submitButton = getByText('forms.buttons.submit.default');
+    expect(isFormValid(submitButton.closest('form'))).to.be.true();
+
+    await new Promise((resolve) => {
+      onSubmit.callsFake(resolve);
+      // eslint-disable-next-line no-restricted-syntax
+      userEvent.click(submitButton);
+    });
+
+    // At this point, the page should redirect, so we do not expect that the user should be prompted
+    // about unsaved changes in navigating.
+    const event = new window.Event('beforeunload', { cancelable: true, bubbles: false });
+    window.dispatchEvent(event);
+    expect(event.defaultPrevented).to.be.false();
+  });
+
+  it('renders unhandled submission failure', async () => {
+    const { getByLabelText, getByText, getAllByText, findAllByText, findByText } = render(
+      <AcuantContextProvider sdkSrc="about:blank" cameraSrc="about:blank">
+        <DocumentCapture />
+      </AcuantContextProvider>,
+      {
+        uploadError: new Error('Server unavailable'),
+        expectedUploads: 2,
+      },
+    );
+
+    let submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.click(submitButton);
+    await findAllByText('simple_form.required.text');
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_front'), validUpload);
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_back'), validUpload);
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+    await userEvent.click(submitButton);
+
+    await findByText('doc_auth.errors.general.network_error');
+
+    expect(console).to.have.loggedError(/^Error: Uncaught/);
+    expect(console).to.have.loggedError(
+      /React will try to recreate this component tree from scratch using the error boundary you provided/,
+    );
+
+    // Make sure that the first element after a tab is what we expect it to be.
+    await userEvent.tab();
+    const firstFocusable = getByLabelText('doc_auth.headings.document_capture_front');
+    expect(document.activeElement).to.equal(firstFocusable);
+
+    const hasValueSelected = getAllByText('doc_auth.forms.change_file').length === 2;
+    expect(hasValueSelected).to.be.true();
+
+    // Verify re-submission. It will fail again, but test can at least assure that the interstitial
+    // screen is shown once more.
+
+    submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.click(submitButton);
+
+    await findByText('doc_auth.errors.general.network_error');
+
+    expect(console).to.have.loggedError(/^Error: Uncaught/);
+    expect(console).to.have.loggedError(
+      /React will try to recreate this component tree from scratch using the error boundary you provided/,
+    );
+  });
+
+  it('renders handled submission failure', async () => {
+    const uploadError = new UploadFormEntriesError();
+    uploadError.formEntryErrors = [
+      { field: 'front', message: 'Image has glare' },
+      { field: 'back', message: 'Please fill in this field' },
+      { message: 'An unknown error occurred' },
+    ].map(toFormEntryError);
+    const { getByLabelText, getByText, getAllByText, findAllByText, findAllByRole } = render(
+      <AcuantContextProvider sdkSrc="about:blank" cameraSrc="about:blank">
+        <DocumentCapture />
+      </AcuantContextProvider>,
+      {
+        uploadError,
+        expectedUploads: 2,
+      },
+    );
+
+    let submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.click(submitButton);
+    await findAllByText('simple_form.required.text');
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_front'), validUpload);
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_back'), validUpload);
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+    await userEvent.click(submitButton);
+
+    let notices = await findAllByRole('alert');
+    expect(notices[0].textContent).to.equal('Image has glare');
+    expect(notices[1].textContent).to.equal('Please fill in this field');
+    expect(getByText('An unknown error occurred')).to.be.ok();
+
+    expect(console).to.have.loggedError(/^Error: Uncaught/);
+    expect(console).to.have.loggedError(
+      /React will try to recreate this component tree from scratch using the error boundary you provided/,
+    );
+
+    // Make sure that the first focusable element after a tab is what we expect it to be.
+    await userEvent.tab();
+    const firstFocusable = getByLabelText('doc_auth.headings.document_capture_front');
+    expect(document.activeElement).to.equal(firstFocusable);
+
+    const hasValueSelected = !!getByLabelText('doc_auth.headings.document_capture_front');
+    expect(hasValueSelected).to.be.true();
+
+    submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_front'), validUpload);
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_back'), validUpload);
+
+    // Once fields are changed, their notices should be cleared. If all field-specific errors are
+    // addressed, submit should be enabled once more.
+    notices = await findAllByRole('alert');
+    const errorNotices = notices.filter((notice) => notice.classList.contains('usa-alert--error'));
+    expect(errorNotices).to.have.lengthOf(0);
+
+    // Verify re-submission. It will fail again, but test can at least assure that the interstitial
+    // screen is shown once more.
+    await userEvent.click(submitButton);
+
+    await waitFor(() => expect(() => getAllByText('doc_auth.info.interstitial_eta')).to.throw());
+
+    expect(console).to.have.loggedError(/^Error: Uncaught/);
+    expect(console).to.have.loggedError(
+      /React will try to recreate this component tree from scratch using the error boundary you provided/,
+    );
+  });
+
   it('redirects from a server error', async () => {
     const endpoint = '/upload';
     const { getByLabelText, getByText } = render(
@@ -117,6 +301,86 @@ describe('document-capture/components/document-capture', () => {
     expect(event.defaultPrevented).to.be.false();
   });
 
+  it('renders async upload pending progress', async () => {
+    const statusChecks = 3;
+    let remainingStatusChecks = statusChecks;
+    sandbox.stub(window, 'fetch').resolves({ ok: true, headers: new window.Headers() });
+    const upload = sinon.stub().callsFake((payload, { endpoint }) => {
+      switch (endpoint) {
+        case 'about:blank#upload':
+          expect(payload).to.have.keys([
+            'front_image_iv',
+            'front_image_url',
+            'front_image_metadata',
+            'back_image_iv',
+            'back_image_url',
+            'back_image_metadata',
+            'flow_path',
+          ]);
+
+          return Promise.resolve({ success: true, isPending: true });
+        case 'about:blank#status':
+          expect(payload).to.be.empty();
+
+          return Promise.resolve({ success: true, isPending: Boolean(remainingStatusChecks--) });
+        default:
+          throw new Error();
+      }
+    });
+    const key = await window.crypto.subtle.generateKey(
+      {
+        name: 'AES-GCM',
+        length: 256,
+      },
+      true,
+      ['encrypt', 'decrypt'],
+    );
+
+    const { getByLabelText, getByText, getAllByText, findAllByText } = baseRender(
+      <UploadContextProvider
+        endpoint="about:blank#upload"
+        statusEndpoint="about:blank#status"
+        statusPollInterval={0}
+        backgroundUploadURLs={{
+          front: 'about:blank#front',
+          back: 'about:blank#back',
+        }}
+        backgroundUploadEncryptKey={key}
+        upload={upload}
+      >
+        <AcuantContextProvider sdkSrc="about:blank" cameraSrc="about:blank">
+          <DocumentCapture isAsyncForm />
+        </AcuantContextProvider>
+      </UploadContextProvider>,
+    );
+
+    const submitButton = getByText('forms.buttons.submit.default');
+    await userEvent.click(submitButton);
+    await findAllByText('simple_form.required.text');
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_front'), validUpload);
+    await userEvent.upload(getByLabelText('doc_auth.headings.document_capture_back'), validUpload);
+    await waitFor(() => expect(() => getAllByText('simple_form.required.text')).to.throw());
+
+    return new Promise((resolve) => {
+      onSubmit.callsFake(() => {
+        // Error logged at initial pending retry.
+        expect(console).to.have.loggedError(/^Error: Uncaught/);
+        expect(console).to.have.loggedError(/React will try to recreate this component/);
+
+        // Error logged at every scheduled check thereafter.
+        for (let i = 0; i < statusChecks; i++) {
+          expect(console).to.have.loggedError(/^Error: Uncaught/);
+          expect(console).to.have.loggedError(/React will try to recreate this component/);
+        }
+
+        resolve();
+      });
+
+      // eslint-disable-next-line no-restricted-syntax
+      userEvent.click(submitButton);
+    });
+  });
+
   describe('pending promise values', () => {
     let completeUploadAsSuccess;
     let completeUploadAsFailure;
```